balancer/randomsubsetting/randomsubsetting_test.go

```diff
@@ -0,0 +1,286 @@
+		server := stubserver.StartTestService(t, nil)
+		t.Cleanup(server.Stop)
+		addresses[i] = resolver.Address{Addr: server.Address}
+	}
+	return addresses
+}
+
+func startTestSetupClients(t *testing.T, clientsCount int, subsetSize int, addresses []resolver.Address) []testgrpc.TestServiceClient {
+	t.Helper()
+
+	clients := make([]testgrpc.TestServiceClient, clientsCount)
+	ccs := make([]*grpc.ClientConn, clientsCount)
+	var err error
+
+	for i := 0; i < clientsCount; i++ {
+		mr := manual.NewBuilderWithScheme("subsetting-e2e")
+		jsonConfig := fmt.Sprintf(`
+		{
+		  "loadBalancingConfig": [
+			{
+			  "random_subsetting": {
+				"subsetSize": %d,
+				"childPolicy": [{"round_robin": {}}]
+			  }
+			}
+		  ]
+		}`, subsetSize)
+
+		sc := internal.ParseServiceConfig.(func(string) *serviceconfig.ParseResult)(string(jsonConfig))
+		mr.InitialState(resolver.State{
+			Addresses:     addresses,
+			ServiceConfig: sc,
+		})
+
+		ccs[i], err = grpc.Dial(mr.Scheme()+":///", grpc.WithResolvers(mr), grpc.WithTransportCredentials(insecure.NewCredentials()))
+		if err != nil {
+			t.Fatalf("grpc.Dial() failed: %v", err)
+		}
+		clients[i] = testgrpc.NewTestServiceClient(ccs[i])
+	}
+
+	t.Cleanup(func() {
+		for _, cc := range ccs {
+			cc.Close()
+		}
+	})
+	return clients
+}
+
+func checkRoundRobinRPCs(ctx context.Context, t *testing.T, clients []testgrpc.TestServiceClient, subsetSize int, maxDiff int) {
+	clientsPerBackend := map[string]map[int]struct{}{}
+
+	for clientIdx, client := range clients {
+		// make sure that every client send exactly 1 request to each server in its subset
+		for i := 0; i < subsetSize; i++ {
+			var peer peer.Peer
+			_, err := client.EmptyCall(ctx, &testpb.Empty{}, grpc.Peer(&peer))
+			if err != nil {
+				t.Fatalf("failed to call server: %v", err)
+			}
+			if peer.Addr != nil {
+				if m, ok := clientsPerBackend[peer.Addr.String()]; !ok {
+					clientsPerBackend[peer.Addr.String()] = map[int]struct{}{clientIdx: {}}
+				} else if _, ok := m[clientIdx]; !ok {
+					m[clientIdx] = struct{}{}
+				} else {
+					// The backend receives a second request from the same client. This could happen if the client have 1 backend in READY
+					// state while the other  are CONNECTING. In this case round_robbin will pick the same address twice.
+					// We are going to retry after short timeout.
+					time.Sleep(10 * time.Microsecond)
+					i--
+				}
+			} else {
+				t.Fatalf("peer.Addr == nil, peer: %v", peer)
+			}
+		}
+	}
+
+	minClientsPerBackend := math.MaxInt
+	maxClientsPerBackend := 0
+	for _, v := range clientsPerBackend {
+		if len(v) < minClientsPerBackend {
+			minClientsPerBackend = len(v)
+		}
+		if len(v) > maxClientsPerBackend {
+			maxClientsPerBackend = len(v)
+		}
+	}
+
+	if maxClientsPerBackend > minClientsPerBackend+maxDiff {
+		t.Fatalf("the difference between min and max clients per backend should be <= %d, clientsPerBackend: %v", maxDiff, clientsPerBackend)
+	}
+}
+
+func (s) TestSubsettingE2E(t *testing.T) {
```

Comment:
The only real piece of functionality provided by this LB policy is the subsetting algorithm. If we refactor it the way I recommended in another comment, then we could simply have unit tests for that method to ensure that subsetting behavior is as expected. 

And then we could have a functional test where we:
- create an instance of the LB policy, 
- call its `UpdateClientConnState` method to pass it some configuration with a child policy that is controlled by the test. See here for how to define a balancer that is completely under the control of the test: https://github.com/grpc/grpc-go/blob/master/internal/balancer/stub/stub.go
- Verify that the child policy gets the expected `resolver.State`

There is very little value is having an E2E style for this LB policy.